### ISSUE TITLE
Setup a production webpack config

### ISSUE THREAD

**Issue by pfiaux** [Cyclodex/cyclone#38]
This is related to #32, when loading the libraries from NPM they'll need to be minified, at least in production.

There's a few things that would improve the dist size of the app:
* Disable sourcemaps or store them outside of the main build file
* Add a minifiying plugin uglify or something else

These 2 should help a lot, if it's generating a sourcemap for all of angular and including it inline 

    new webpack.optimize.CommonsChunkPlugin('common.js'),
    new webpack.optimize.DedupePlugin(),
    new webpack.optimize.UglifyJsPlugin(),
    new webpack.optimize.AggressiveMergingPlugin()

Here's some resources I found quickly looking into it:
* https://medium.com/@mrbar42/how-i-reduced-the-size-of-my-webpack-bundle-by-1000-f4d74894c2e5#.per1k45iu
* http://stackoverflow.com/a/35055803
* https://webpack.github.io/docs/optimization.html
* https://webpack.github.io/docs/list-of-plugins.html#optimize

- [x] Setup a production config & NPM Scripts
- [x] Remove sourcemaps
- [x] Add optimization plugins

**Comment by Cyclodex** [comment]
**Production Configuration Update #2: Node Environment**
I also checked if the production variable would make it different, but the generated size is still the same. So not committing these changes. But for react it would make a difference.

```javascript
plugins: [
  new webpack.DefinePlugin({
    'process.env': {
      'NODE_ENV': JSON.stringify('production')
    }
  })
],
```

**Comment by Cyclodex** [comment]
When building it works, when I build with the `-p` function to the angular throws an error:
`(index):47959 Uncaught Error: [$injector:modulerr] Failed to instantiate module cycloneApp due to:
Error: [$injector:unpr] Unknown provider: e`
- https://docs.angularjs.org/error/$injector/unpr?p0=e

Any idea? I think it could be something we have missing, perhaps something with the requires? Hmm.

**Comment by Cyclodex** [comment]
Good page with webpack and angular guidance:

- https://github.com/dmachat/angular-webpack-cookbook/wiki

**Comment by pfiaux** [comment]
I'm not sure if you need the NODE ENV var, to set it you'd need to prefix the build command with `NODE_ENV=production weback.....`

The unknown provider could come from missing annotation when minified, like if you have `service('blahService`, ['$q', function ($q) {...}]` that works but if you have just service('blahService`,  function ($q) {...}` and $q gets renamed to e because of uglify then it's broken. You could set up [ng-annotate](https://github.com/jeffling/ng-annotate-webpack-plugin) which should do it automatically, or make sure you use the long form with the string describing the parameters.

Also if you look at the [uglify plugin](https://webpack.github.io/docs/list-of-plugins.html#uglifyjsplugin) there's an option to tell it to not mangle global stuff so you probably want to configure it like:

    new webpack.optimize.UglifyJsPlugin({
      mangle: {
        except: ['angular'] // maybe also firebase? not sure
      }
    })

Then should be able to still mangle some code.

**Comment by Cyclodex** [comment]
Okey I think I got these things done.
**NODE_ENV:**
I prefer to have this within the plugin, instead on build command. I think it doesn't matter which way one would do it.

**ng-annotate**
Add this and it fixed the issue, so yes it seems that somewhere is something missing. I will check if I see where the long form is missing.

**mangle**
Yes I already looked into the mangle, and disabled it before I had the ng-annotate to have it working.
Now its enabled again 👍 

Thanks for your support, feel free to add any comment, feedback or further review the code.

**Next**
Next we should probably put all the production code into a separate folder like `dist` or something.
However this will also need that we take the html and other assets into the webpack.
Will try to look into that at some point.
